agenthub/langchains_agent/__init__.py

```diff
@@ -43,59 +63,135 @@
-                    event = Event("run", {"command": command})
+                    d = {"action": "run", "args": {"command": command}}
                     next_is_output = True
+
                 elif thought.startswith("RECALL"):
                     query = thought.split("RECALL ")[1]
-                    event = Event("recall", {"query": query})
+                    d = {"action": "recall", "args": {"query": query}}
                     next_is_output = True
+
                 elif thought.startswith("BROWSE"):
                     url = thought.split("BROWSE ")[1]
-                    event = Event("browse", {"url": url})
+                    d = {"action": "browse", "args": {"url": url}}
                     next_is_output = True
                 else:
-                    event = Event("think", {"thought": thought})
-            self.agent.add_event(event)
-        self._initialized = True
+                    d = {"action": "think", "args": {"thought": thought}}
 
-    def add_event(self, event: Event) -> None:
-        if self.agent:
-            self.agent.add_event(event)
+        self._add_event(d)
+        self._initialized = True
 
-    def step(self, cmd_mgr) -> Event:
+    def step(self, state: State) -> Action:
         self._initialize()
-        return self.agent.get_next_action(cmd_mgr)
+        # TODO: make langchains agent use Action & Observation
+        # completly from ground up
 
-    def search_memory(self, query: str) -> List[str]:
-        return self.agent.memory.search(query)
+        # Translate state to action_dict
+        for info in state.updated_info:
+            if isinstance(info, Observation):
+                if isinstance(info, CmdOutputObservation):
+                    if info.error:
+                        d = {"action": "error", "args": {"output": info.content}}
+                    else:
+                        d = {"action": "output", "args": {"output": info.content}}
+                # elif isinstance(info, UserMessageObservation):
+                #     d = {"action": "output", "args": {"output": info.message}}
+                # elif isinstance(info, AgentMessageObservation):
+                #     d = {"action": "output", "args": {"output": info.message}}
+                elif isinstance(info, BrowserOutputObservation):
+                    d = {"action": "output", "args": {"output": info.content}}
+                else:
+                    raise NotImplementedError(f"Unknown observation type: {info}")
+                self._add_event(d)
+            elif isinstance(info, Action):
+                if isinstance(info, CmdRunAction):
+                    d = {"action": "run", "args": {"command": info.command}}
+                elif isinstance(info, CmdKillAction):
+                    d = {"action": "kill", "args": {"id": info.id}}
+                elif isinstance(info, BrowseURLAction):
+                    d = {"action": "browse", "args": {"url": info.url}}
```

Comment:
I think so! But I feel we should probably prioritize getting this abstraction merged, then start a separate PR to re-write the inner workings of `langchains` agent completely to adopt this `Action` and `Observation` (and the serialization via `to_dict`) -- otherwise it will be pretty challenge for me to keep chasing with all the new commits keeping pushing to `main` everyday hahah